packages/angular/ssr/src/i18n.ts

```diff
@@ -43,3 +43,147 @@ export function getPotentialLocaleIdFromUrl(url: URL, basePath: string): string
+ * - Quality value handling (e.g., `q=1`, `q=0.8`). If no quality value is provided, it defaults to `q=1`.
+ * - Prefix matching (e.g., `en` matching `en-US` or `en-GB`).
+ *
+ * @param header - The `Accept-Language` header string to parse and evaluate. It may contain multiple
+ *                 locales with optional quality values, for example: `'en-US;q=0.8,fr-FR;q=0.9'`.
+ * @param supportedLocales - An array of supported locales (e.g., `['en-US', 'fr-FR']`),
+ *                           representing the locales available in the application.
+ * @returns The best matching locale from the supported languages, or `undefined` if no match is found.
+ *
+ * @example
+ * ```js
+ * getPreferredLocale('en-US;q=0.8,fr-FR;q=0.9', ['en-US', 'fr-FR', 'de-DE'])
+ * // returns 'fr-FR'
+ *
+ * getPreferredLocale('en;q=0.9,fr-FR;q=0.8', ['en-US', 'fr-FR', 'de-DE'])
+ * // returns 'en-US'
+ *
+ * getPreferredLocale('es-ES;q=0.7', ['en-US', 'fr-FR', 'de-DE'])
+ * // returns undefined
+ * ```
+ */
+export function getPreferredLocale(
+  header: string,
+  supportedLocales: ReadonlyArray<string>,
+): string | undefined {
+  const parsedLocales = parseLanguageHeader(header);
+
+  // Handle edge cases:
+  // - No preferred locales provided.
+  // - Only one supported locale.
+  // - Wildcard preference.
+  if (
+    parsedLocales.size === 0 ||
+    supportedLocales.length < 2 ||
+    (parsedLocales.size === 1 && parsedLocales.has('*'))
+  ) {
+    return supportedLocales[0];
+  }
+
+  // Create a map for case-insensitive lookup of supported locales.
+  // Keys are normalized (lowercase) locale values, values are original casing.
+  const normalizedSupportedLocales = new Map<string, string>();
+  for (const locale of supportedLocales) {
+    normalizedSupportedLocales.set(locale.toLowerCase(), locale);
+  }
+
+  // Iterate through parsed locales in descending order of quality.
+  let bestMatch: string | undefined;
+  const qualityZeroNormalizedLocales = new Set<string>();
+  for (const [locale, quality] of parsedLocales) {
+    const normalizedLocale = locale.toLowerCase();
+    if (quality === 0) {
+      qualityZeroNormalizedLocales.add(normalizedLocale);
+      continue; // Skip locales with quality value of 0.
+    }
+
+    // Exact match found.
+    if (normalizedSupportedLocales.has(normalizedLocale)) {
+      return normalizedSupportedLocales.get(normalizedLocale);
+    }
+
+    // If an exact match is not found, try prefix matching (e.g., "en" matches "en-US").
+    // Store the first prefix match encountered, as it has the highest quality value.
+    if (bestMatch !== undefined) {
+      continue;
+    }
+
+    const [languagePrefix] = normalizedLocale.split('-', 1);
+    for (const supportedLocale of normalizedSupportedLocales.keys()) {
+      if (supportedLocale.startsWith(languagePrefix)) {
+        bestMatch = normalizedSupportedLocales.get(supportedLocale);
+        break; // No need to continue searching for this locale.
+      }
+    }
```

Comment:
Perf: This is a `O(parsedLocales.length * normalizedSupportedLocales.size)` check. I wonder if we could do something slightly more optimal by using a tree instead of a map, where each node in the tree is a locale segment. Such that you could have:

```
en -> US
   |
    > GB
it
es
```

Then looking up a locale is just walking the tree and if we don't find it, we use the deepest result we found which would inherently be the most specific locale.

If we limit this implementation to just two locale segments, it's probably not that important and likely fine to ignore for now. If we wanted to generically support N locale segments, it might be a useful optimization.

Reply:
I'm not overly concerned about performance in this case, as the number of `parsedLocales` and `normalizedSupportedLocales` is usually quite small.

We could definitely consider this in the future.